Python/ceval.c

```diff
@@ -1061,34 +1062,133 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+    if (*pexecutor != NULL) {
+#ifdef Py_DEBUG
+        PyCodeObject *code = _PyFrame_GetCode(frame);
+        DPRINTF(2, "Jumping to new executor for %s (%s:%d) at byte offset %d\n",
+                PyUnicode_AsUTF8(code->co_qualname),
+                PyUnicode_AsUTF8(code->co_filename),
+                code->co_firstlineno,
+                2 * (int)(frame->instr_ptr - _PyCode_CODE(_PyFrame_GetCode(frame))));
+#endif
+        Py_DECREF(current_executor);
+        current_executor = (_PyUOpExecutorObject *)*pexecutor;
+        Py_INCREF(current_executor);
+        goto enter_tier_two;
+    }
+
+    // Increment and check side exit counter.
+    // (Even though we only need it for certain opcodes.)
     next_instr = frame->instr_ptr;
-    goto resume_frame;
+    uint16_t *pcounter = current_executor->counters + pc;
+    *pcounter += 1 << OPTIMIZER_BITS_IN_COUNTER;
+    /* We are using unsigned values, but we really want signed values, so
+     * do the 2s complement comparison manually */
+    uint16_t ucounter = *pcounter + (1 << 15);
+    uint16_t threshold = tstate->interp->optimizer_resume_threshold + (1 << 15);
+    if (ucounter <= threshold)
+    {
+        Py_DECREF(current_executor);
+        goto resume_frame_using_stack_pointer;
+    }
+
+    // Decode instruction to look past EXTENDED_ARG.
+    opcode = next_instr[0].op.code;
+    if (opcode == EXTENDED_ARG) {
+        opcode = next_instr[1].op.code;
+    }
+
+    // For selected opcodes build a new executor and enter it now.
+    if (opcode == POP_JUMP_IF_FALSE ||
+        opcode == POP_JUMP_IF_TRUE ||
+        opcode == POP_JUMP_IF_NONE ||
+        opcode == POP_JUMP_IF_NOT_NONE)
+    {
+        DPRINTF(2, "--> %s @ %d in %p has %d side exits\n",
+                _PyUOpName(uopcode), pc, current_executor, (int)(*pcounter));
+        DPRINTF(2, "    T1: %s\n", _PyOpcode_OpName[opcode]);
+
+        int optimized = _PyOptimizer_Unanchored(frame, next_instr, pexecutor, stack_pointer);
+        if (optimized < 0) {
+            goto error_tier_two;
+        }
+
+        if (!optimized) {
+            DPRINTF(2, "--> Failed to optimize %s @ %d in %p\n",
+                    _PyUOpName(uopcode), pc, current_executor);
+        }
+        else {
+#ifdef Py_DEBUG
+            DPRINTF(1, "--> Optimized %s @ %d in %p\n",
+                    _PyUOpName(uopcode), pc, current_executor);
+            PyCodeObject *code = _PyFrame_GetCode(frame);
+            DPRINTF(2, "Jumping to fresh executor for %s (%s:%d) at byte offset %d\n",
+                    PyUnicode_AsUTF8(code->co_qualname),
+                    PyUnicode_AsUTF8(code->co_filename),
+                    code->co_firstlineno,
+                    2 * (int)(frame->instr_ptr - _PyCode_CODE(_PyFrame_GetCode(frame))));
+#endif
+            Py_DECREF(current_executor);
+            current_executor = (_PyUOpExecutorObject *)*pexecutor;
+
+            // Reject trace if it repeats the uop that just deoptimized.
```

Comment:
This test may be a bit imprecise(*), but it tries to discard the case where, even though the counter in the executor indicated that this side exit is "hot", the Tier 1 bytecode hasn't been re-specialized yet. In that case the trace projection will just repeat the uop that just took a deopt side exit, causing it to immediately deopt again. This seems a waste of time and executors -- eventually the sub-executor's deopt counter will also indicate it is hot, and then we'll try again, but it seems better (if we can catch it) to avoid creating the sub-executor in the first place, relying on exponential backoff for the side-exit counter instead (implemented below at L1180 and ff.).

For various reasons, the side-exit counters and the Tier 1 deopt counters don't run in sync, so it's possible that the side-exit counter triggers before the Tier 1 counter has re-specialized. This check gives that another chance.

The test that I would *like* to use here would be check if the Tier 1 opcode is still unchanged (i.e., not re-specialized), but the executor doesn't record that information (and it would take up a lot of space, we'd need an extra byte for each uop that can deoptimize at least).

___
(*) The test I wrote is exact for the conditional branches I special-cased above (that's why there's a further special case here for `_IS_NONE`). For other opcodes it may miss a few cases, e.g. when a single T1 bytecode translates to multiple guards and the failing guard is not the first uop in the translation (i.e. this would always happens for calls, whose translation always starts with `_PEP_523`, which never dopts in cases we care about). In those cases we can produce a sub-executor that immediately deoptimizes. (And we never try to re-create executors, no matter how often it deoptimizes -- that's a general flaw in the current executor architecture that we should probably file separately.)